src/main/java/edu/cornell/kfs/pdp/batch/service/impl/ProcessPdpCancelPaidHelperServiceImpl.java

```diff
@@ -0,0 +1,149 @@
+            PaymentSourceToExtractService<PaymentSource> extractService = getPaymentSourceToExtractService(paymentDetail);
+            if (extractService != null) {
+                try {
+                    PaymentSource dv = (PaymentSource) documentService.getByDocumentHeaderId(documentNumber);
+                    if (dv != null) {
+                        if (disbursedPayment || primaryCancel || crCancel) {
+                            if (!crCancel) {
+                                extractService.cancelPayment(dv, processDate);
+                            }
+                        } else {
+                            extractService.resetFromExtraction(dv);
+                        }
+                    }
+                } catch (WorkflowException we) {
+                    throw new RuntimeException("Could not retrieve document #" + documentNumber, we);
+                }
+            } else {
+                LOG.warn("processPdpCancel() Unknown document type (" + documentTypeCode + ") for document ID: " + documentNumber);
+                return;
+            }
+        }
+
+        paymentGroupService.processCancelledGroup(paymentDetail.getPaymentGroup(), processDate);
+    }
+
+    /**
+     * Default implementation uses most of the "while" loop contents from the ProcessPdpCancelPaidServiceImpl.processPdpPaids method,
+     * with additional CU-related changes as needed.
+     */
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    @Override
+    public void processPdpPaid(PaymentDetail paymentDetail, Date processDate) {
+        String documentTypeCode = paymentDetail.getFinancialDocumentTypeCode();
+        String documentNumber = paymentDetail.getCustPaymentDocNbr();
+
+        if (purchasingAccountsPayableModuleService.isPurchasingBatchDocument(documentTypeCode)) {
+            purchasingAccountsPayableModuleService.handlePurchasingBatchPaids(documentNumber, documentTypeCode, processDate);
+        } else {
+            PaymentSourceToExtractService<PaymentSource> extractService = getPaymentSourceToExtractService(paymentDetail);
+            if (extractService != null) {
+                try {
+                    PaymentSource dv = (PaymentSource) documentService.getByDocumentHeaderId(documentNumber);
+                    extractService.markAsPaid(dv, processDate);
+                } catch (WorkflowException we) {
+                    throw new RuntimeException("Could not retrieve document #" + documentNumber, we);
+                }
+            } else {
+                LOG.warn("processPdpPaids() Unknown document type (" + documentTypeCode + ") for document ID: " + documentNumber);
+                return;
+            }
+        }
+
+        paymentGroupService.processPaidGroup(paymentDetail.getPaymentGroup(), processDate);
+    }
+
+    // Copied from ProcessPdpCancelPaidServiceImpl
+    protected PaymentSourceToExtractService<PaymentSource> getPaymentSourceToExtractService(PaymentDetail paymentDetail) {

```

Comment:
This will make upgrades difficult. If anything changes in the base method in ProcessPdpCancelPaidServiceImpl that this was copied from we may miss those changes.

Reply:
Ok, I will see if I can refactor the code to avoid as much of this duplication as possible, while still making the new transaction management possible.